### ISSUE TITLE
Type error with ppx and camlp4 syntax extensions

### ISSUE THREAD

**Issue by vouillon** [ocsigen/eliom#234]
This does not typecheck:
````
let x = 1
let%client _ : int = ~%x
let x = "abcd"
let%client _ : string = ~%x
````
Indeed, the type of the second instance of `x` overrides the type of the first one:
````
$ js_of_eliom  -ppx -c  toto.eliom
File "toto.eliom", line 2, characters 23-24:
Error: This expression has type string but an expression was expected of type int
````
We have the same kind of issue with the camlp4 syntax extension, though in this case, the type error is on the second instance of `x`.
```
let x = 1
{client{ let _ : int = %x }}
let x = "abcd"
{client{ let _ : string = %x }}
```

**Comment by vasilisp** [comment]
I think we build a hashtable out of all the identifiers in the `type_mli`, with possible shadowing. We probably need to add locations in there, and do lookup that respects the scope.

**Comment by Drup** [comment]
Yes, it should be enough to add an hash of the position to the identifier when we build the .mli

**Comment by vouillon** [comment]
Or we could empty the list `injected_idents` in `ppx_eliom_utils.ml` when leaving a client or shared section.

**Comment by vasilisp** [comment]
I hope that 67d651e fixes the issue for PPX (with @vouillon's approach).

**Comment by vasilisp** [comment]
Fixed via 17086b9 and 22ef439.

**Comment by Breadcorn** [comment]
> This does not typecheck:
> 
> ```
> let x = 1
> let%client _ : int = ~%x
> let x = "abcd"
> let%client _ : string = ~%x
> ```
> 
> Indeed, the type of the second instance of `x` overrides the type of the first one:
> 
> ```
> $ js_of_eliom  -ppx -c  toto.eliom
> File "toto.eliom", line 2, characters 23-24:
> Error: This expression has type string but an expression was expected of type int
> ```
> 
> We have the same kind of issue with the camlp4 syntax extension, though in this case, the type error is on the second instance of `x`.
> 
> ```
> let x = 1
> {client{ let _ : int = %x }}
> let x = "abcd"
> {client{ let _ : string = %x }}
> ```

Gix
